Client/Frontend/Browser/OpenPdfHelper.swift

```diff
@@ -0,0 +1,150 @@
+    static func canOpen(url: NSURL) -> Bool
+    func showOpenInView(inView parentView: UIView, forWebView webView: WKWebView)
+    func hideOpenInView(forWebView webView: WKWebView)
+    func open()
+}
+
+struct OpenInHelperFactory {
+    static func helperForURL(url: NSURL) -> OpenInHelper? {
+        if OpenPdfInHelper.canOpen(url) {
+            return OpenPdfInHelper(url: url)
+        }
+
+        return nil
+    }
+}
+
+class OpenPdfInHelper: NSObject, OpenInHelper, UIDocumentInteractionControllerDelegate {
+    private var view: OpenInView?
+    private var docController: UIDocumentInteractionController? = nil
+
+    private var parentView: UIView!
+    private var webView: WKWebView!
+
+    init(url: NSURL) {
+        super.init()
+        let contentsOfFile = NSData(contentsOfURL: url)
+        let dirPaths = NSSearchPathForDirectoriesInDomains(.DocumentDirectory,
+            .UserDomainMask, true)
+
+        let dirPath = NSURL(fileURLWithPath: dirPaths[0]).URLByAppendingPathComponent("pdfs")
+        let fileManager = NSFileManager.defaultManager()
+        do {
+            guard let lastPathComponent = url.lastPathComponent else {
+                log.error("failed to create proper URL")
+                return
+            }
+            let filePath = dirPath.URLByAppendingPathComponent(lastPathComponent)
+            try fileManager.createDirectoryAtURL(dirPath, withIntermediateDirectories: true, attributes: nil)
+            if fileManager.createFileAtPath(filePath.absoluteString, contents: contentsOfFile, attributes: nil) {
+                let openInURL = NSURL(fileURLWithPath: filePath.absoluteString)
+                docController = UIDocumentInteractionController(URL: openInURL)
+                docController?.delegate = self
+            } else {
+                log.error("Unable to create local version of PDF file at \(filePath)")
+            }
+        } catch {
+            log.error("Error on creating directory at \(dirPath)")
+        }
+    }
+    
+    static func canOpen(url: NSURL) -> Bool {
+        guard let pathExtension = url.pathExtension else { return false }
+        return pathExtension == FileType.PDF.rawValue && UIApplication.sharedApplication().canOpenURL(NSURL(string: "itms-books:")!)
+    }
+
+    func showOpenInView(inView parentView: UIView, forWebView webView: WKWebView) {
+        let overlayView = OpenInView()
+
+        overlayView.openInButton.addTarget(self, action: "open", forControlEvents: .TouchUpInside)
+        parentView.addSubview(overlayView)
+        overlayView.snp_makeConstraints { make in
+            make.trailing.equalTo(parentView)
+            make.leading.equalTo(parentView)
+            make.height.equalTo(OpenInViewUX.ViewHeight)
+            make.bottom.equalTo(webView.snp_top)
+        }
+
+        webView.snp_updateConstraints { make in
+            make.edges.equalTo(EdgeInsetsMake(OpenInViewUX.ViewHeight, left: 0, bottom: 0, right: 0))
+        }
```

Comment:
The decision was mostly based on the fact that `BrowserViewController` is already over 2000 lines long and confusing as hell and I wanted to not add to it. I felt that keeping things self contained with all responsibility for creation, display and action within one helper would make it easier to maintain in future rather than losing it all in the vastness of BVC.

However, the toolbar solution works just as well. Will explore that and see where it goes.

To be honest, I just hadn't tried the tab tray when testing so didn't even think about it. Stupid really.